agent/src/capability/configstore.cpp

```diff
@@ -0,0 +1,237 @@
+            log_error() << "Capability is not a json object";
+            return ReturnCode::FAILURE;
+        }
+
+        std::string capName;
+        if (!JSONParser::read(capability,"name",capName)) {
+            log_error() << "Could not read Capability name";
+            return ReturnCode::FAILURE;
+        }
+
+        json_t *gateways = json_object_get(capability, "gateways");
+        if (gateways == nullptr) {
+            log_error() << "Could not read Gateways";
+            return ReturnCode::FAILURE;
+        }
+        if (!json_is_array(gateways)) {
+            log_error() << "Gateways is not an array";
+            return ReturnCode::FAILURE;
+        }
+
+        return parseGateways(capName,gateways);
+    }
+    return ReturnCode::SUCCESS;
+}
+
+ReturnCode ConfigStore::parseGateways(std::string capName, json_t *gateways)
+{
+    size_t j;
+    json_t *gateway;
+
+    GatewayConfiguration gwConfs;
+
+    json_array_foreach(gateways, j, gateway) {
+        if (!json_is_object(gateway)) {
+            log_error() << "Gateway is not a json object";
+            return ReturnCode::FAILURE;
+        }
+
+        std::string gwID;
+        if (!JSONParser::read(gateway,"id",gwID)) {
+            log_error() << "Could not read Gateway ID";
+            return ReturnCode::FAILURE;
+        }
+
+        json_t *confs = json_object_get(gateway,"config");
+        if (confs == nullptr) {
+            log_error() << "Could not read Gateway Config";
+            return ReturnCode::FAILURE;
+        }
+        if (!json_is_array(confs)) {
+            log_error() << "Gateway Config is not an array";
+            return ReturnCode::FAILURE;
+        }
+        gwConfs[gwID] = confs;
+
+    }
+
+    GatewayConfiguration oldGwConfs = m_configStore[capName];
+
+    /* If there is already a Gateway Configuration saved with the
+     * same ID, then append the configs from this file to the
+     * existing Gateway in the m_configstore
+     */
+    for (auto it=gwConfs.begin(); it!=gwConfs.end(); ++it) {
+        std::string gwID = it->first;
+        if (oldGwConfs.count(gwID) == 1) {
+            json_t *val = oldGwConfs[gwID];
+            if (json_array_extend(val, it->second) == -1) {
+                log_error() << "Could not add Gateway Config to json array: " << gwID;
+                return ReturnCode::FAILURE;
+            }
+        } else {
+            oldGwConfs[gwID] = it->second;
+        }
+    }
+    m_configStore[capName] = oldGwConfs;
+
+    return ReturnCode::SUCCESS;
+}
+
+std::vector<std::string> ConfigStore::fileList(const std::string &filePath)
+{
+    std::vector<std::string> files;
+
+    DIR *dirFile = opendir(filePath.c_str());
+    if (dirFile == nullptr) {
+        log_error() << "Could not open directory: " << filePath << "\n";
+    } else {
+        struct dirent *hFile;
+        while ((hFile = readdir(dirFile)) != nullptr) {
+            // TODO (thenor): we should check for '.' and '..'
```

Comment:
This was in part a reminder about looking in to how to handle ".." or ".". How do you think they should be handled?
I think other checks will make this fail or not fail anyway. Was just thinking if I should handle this as "/" where I don't allow it.